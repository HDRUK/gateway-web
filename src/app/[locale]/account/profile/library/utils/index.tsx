import { IconButton } from "@mui/material";
import { createColumnHelper } from "@tanstack/react-table";
import { LibraryListItem, SelectedLibrary } from "@/interfaces/Library";
import Link from "@/components/Link";
import StyledCheckbox from "@/components/StyledCheckbox";
import TooltipIcon from "@/components/TooltipIcon";
import { CheckIcon, DeleteForeverIcon } from "@/consts/icons";
import { RouteName } from "@/consts/routeName";

const columnHelper = createColumnHelper<LibraryListItem>();

const getColumns = ({
    handleSelect,
    handleRemove,
    selected,
    translations,
}: {
    handleSelect: (data: SelectedLibrary) => void;
    handleRemove: (id: string | number) => void;
    selected: SelectedLibrary;
    translations: { [id: string]: string };
}) => [
    columnHelper.display({
        id: "actions",
        meta: { isPinned: true },
        cell: ({ row }) => {
            const {
                id,
                datasetId,
                name,
                dataCustodianId,
                dataCustodian,
                darEnabled,
                darTemplatePublished,
                cohortEnabled,
            } = row.original;
            return (
                <div style={{ textAlign: "center" }}>
                    <StyledCheckbox
                        checked={selected[id]?.selected}
                        onChange={(_e, value) =>
                            handleSelect({
                                [id]: {
                                    selected: value,
                                    datasetId,
                                    name,
                                    teamId: dataCustodianId,
                                    teamName: dataCustodian,
                                    darEnabled,
                                    darTemplatePublished,
                                    cohortEnabled,
                                },
                            })
                        }
                        size="large"
                        sx={{ p: 0 }}
                        iconSx={{ mr: 0 }}
                    />
                </div>
            );
        },
        header: () => null,
        size: 43,
    }),

    columnHelper.display({
        id: "name",
        cell: ({ row: { original } }) => {
            const { datasetId, name } = original;
            const linkHref = `/${RouteName.DATASET_ITEM}/${datasetId}`;

            return (
                <Link href={linkHref}>
                    <div style={{ textAlign: "center" }}>{name}</div>
                </Link>
            );
        },
        header: () => <span>{translations.name}</span>,
        size: 150,
    }),

    columnHelper.display({
        id: "darEnabled",
        cell: ({
            row: {
                original: { darEnabled },
            },
        }) => (
            <div style={{ textAlign: "center" }}>
                {darEnabled ? <CheckIcon color="primary" /> : "-"}
            </div>
        ),
        header: () => (
            <TooltipIcon
                buttonSx={{ p: 0 }}
                size="small"
                label={translations.darEnabled}
                content={translations.darEnabled}
            />
        ),
        size: 100,
    }),

    columnHelper.display({
        id: "cohortEnabled",
        cell: ({
            row: {
                original: { cohortEnabled },
            },
        }) => (
            <div style={{ textAlign: "center" }}>
                {cohortEnabled ? <CheckIcon color="primary" /> : "-"}
            </div>
        ),
        header: () => <span>{translations.cohortEnabled}</span>,
        size: 100,
    }),

    columnHelper.display({
        id: "dataCustodian",
        cell: ({
            row: {
                original: { dataCustodian },
            },
        }) => <div style={{ textAlign: "center" }}>{dataCustodian}</div>,
        header: () => <span>{translations.dataCustodian}</span>,
        size: 100,
    }),

    columnHelper.display({
        id: "entityType",
        cell: ({
            row: {
                original: { entityType },
            },
        }) => <div style={{ textAlign: "center" }}>{entityType}</div>,
        header: () => <span>{translations.entityType}</span>,
        size: 100,
    }),

    columnHelper.display({
        id: "delete",
        cell: ({ row: { original } }) => {
            return (
                <div style={{ textAlign: "center" }}>
                    <IconButton onClick={() => handleRemove(original.id)}>
                        <DeleteForeverIcon color="primary" />
                    </IconButton>
                </div>
            );
        },
        header: () => null,
        size: 43,
    }),
];

export { getColumns };
