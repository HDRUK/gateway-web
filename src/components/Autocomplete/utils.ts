import { ValueType } from "./Autocomplete";

const getChipLabel = (
    options: { value: string | number; label: string }[],
    value: ValueType
) => {
    if (typeof value === "string") {
        return value;
    }

    return options.find(option => option.value === value)?.label;
};

export { getChipLabel };
