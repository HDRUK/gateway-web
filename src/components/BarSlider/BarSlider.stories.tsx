import type { Meta, StoryObj } from "@storybook/react";
import BarSlider from "./BarSlider";

const mockData = [
    { xValue: [0, 1000.0], yValue: 3 },
    { xValue: [1000.0, 2000.0], yValue: 24 },
    { xValue: [2000.0, 3000.0], yValue: 43 },
    { xValue: [3000.0, 4000.0], yValue: 34 },
    { xValue: [4000.0, 5000.0], yValue: 7 },
    { xValue: [5000.0, 6000.0], yValue: 62 },
    { xValue: [6000.0, 7000.0], yValue: 3 },
    { xValue: [7000.0, 8000.0], yValue: 93 },
    { xValue: [8000.0, 9000.0], yValue: 23 },
    { xValue: [9000.0, 10000.0], yValue: 63 },
];

const meta: Meta<typeof BarSlider> = {
    component: BarSlider,
    tags: ["autodocs"],
};

export default meta;

type Story = StoryObj<typeof BarSlider>;

export const Default: Story = {
    args: {
        step: 1000,
        data: mockData,
        ariaLabel: "Dataset Population",
        height: 130,
        handleRangeChange: newRange => console.log(newRange),
        handleBarClick: clickedBar => console.log(clickedBar),
    },
};
