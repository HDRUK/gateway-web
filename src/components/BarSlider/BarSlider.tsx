"use client";

import React, { useEffect, useMemo } from "react";
import Slider from "@mui/material/Slider";
import { Group } from "@visx/group";
import { ParentSize } from "@visx/responsive";
import { scaleBand, scaleLinear } from "@visx/scale";
import { Bar } from "@visx/shape";
import theme from "@/config/theme";

type dataType = { yValue: number; xValue: number[] };

interface BarsProps {
    width: number;
    height: number;
    selected: number[];
    data: dataType[];
    handleEvent?: (d: dataType) => void;
}

function valuetext(value: number) {
    return `${value}Â°C`;
}

const BarSlider = ({
    data,
    width,
    height,
    selected,
    handleEvent,
}: BarsProps) => {
    // bounds
    const xMax = width;
    const yMax = height;

    // scales, memoize for performance
    const xScale = useMemo(
        () =>
            scaleBand<number>({
                range: [0, xMax],
                round: true,
                domain: data.map(d => d.xValue[0]),
                padding: 0.1,
            }).paddingOuter(0),
        [data, xMax]
    );
    const yScale = useMemo(
        () =>
            scaleLinear<number>({
                range: [yMax, 0],
                round: true,
                domain: [0, Math.max(...data.map(d => d.yValue))],
            }),
        [data, yMax]
    );

    const getIsInRange = (d: dataType) => {
        const [lowerRange, upperRange] = selected;
        const lowerRangeValue = data[lowerRange]?.xValue[0];
        const upperRangeValue = data[upperRange - 1]?.xValue[1];

        return d.xValue[0] >= lowerRangeValue && d.xValue[1] <= upperRangeValue;
    };

    const getBarData = (d: dataType) => {
        const barWidth = xScale.bandwidth();
        const barHeight = yMax - (yScale(d.yValue) ?? 0);
        const barX = xScale(d.xValue[0]);
        const barY = yMax - barHeight;
        return { barWidth, barHeight, barX, barY };
    };

    return width < 10 ? null : (
        <svg width={width} height={height}>
            <rect width={width} height={height} fill="transparent" rx={14} />
            <Group top={0}>
                {data.map(d => {
                    const isInRange = getIsInRange(d);
                    const { barWidth, barHeight, barX, barY } = getBarData(d);
                    return (
                        <Bar
                            key={`bar-${d.xValue[0]}`}
                            x={barX}
                            y={barY}
                            opacity={isInRange ? 1 : 0.5}
                            width={barWidth}
                            height={barHeight}
                            fill={theme.palette.primary.main}
                            onClick={() => {
                                if (typeof handleEvent === "function") {
                                    handleEvent(d);
                                }
                            }}
                        />
                    );
                })}
            </Group>
        </svg>
    );
};

interface BarSliderContainerProps {
    height?: number;
    step: number;
    leftStartPoint?: number;
    rightStartPoint?: number;
    data: dataType[];
    handleRangeChange?: (range: number[] | number) => void;
    handleBarClick?: (data: dataType) => void;
    ariaLabel: string;
    sliderSize?: "small" | "medium";
}

const BarSliderContainer = ({
    height = 400,
    sliderSize = "small",
    leftStartPoint = 0,
    data,
    step,
    handleBarClick,
    handleRangeChange,
    ariaLabel,
    rightStartPoint,
}: BarSliderContainerProps) => {
    const [selected, setSelected] = React.useState<number[]>([
        leftStartPoint,
        rightStartPoint || data.length,
    ]);

    useEffect(() => {
        setSelected([leftStartPoint, rightStartPoint || data.length]);
    }, [data.length, leftStartPoint, rightStartPoint]);

    const onChange = (_e: Event, newValue: number[] | number) => {
        setSelected(newValue as number[]);
        if (typeof handleRangeChange === "function") {
            handleRangeChange(newValue);
        }
    };

    const onEvent = (d: dataType) => {
        if (typeof handleBarClick === "function") {
            handleBarClick(d);
        }
    };

    const valueLabelFormat = (value: number) => {
        const [, high] = data[value - 1]?.xValue || [1, 1];
        return high;
    };
    return (
        <>
            <div style={{ height, marginBottom: "-13px" }}>
                <ParentSize>
                    {({ width, height }) => (
                        <BarSlider
                            handleEvent={onEvent}
                            width={width}
                            selected={selected}
                            data={data}
                            height={height}
                        />
                    )}
                </ParentSize>
            </div>
            <Slider
                sx={{ marginBottom: "-13px" }}
                step={step}
                marks
                getAriaLabel={() => ariaLabel || ""}
                value={selected}
                size={sliderSize}
                max={data.length}
                onChange={onChange}
                valueLabelDisplay="auto"
                valueLabelFormat={valueLabelFormat}
                getAriaValueText={valuetext}
            />
        </>
    );
};

export default BarSliderContainer;
