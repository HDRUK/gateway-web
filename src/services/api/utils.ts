import { NotificationOptions } from "@/interfaces/Api";
import { AxiosResponse } from "axios";
import { Error } from "@/interfaces/Error";
import notificationService from "../notification";

interface ErrorNotificationProps {
    props: NotificationOptions;
    errorResponse: AxiosResponse<Error>;
    method: "delete" | "post" | "put" | "get";
}

const errorNotification = ({
    errorResponse,
    method,
    props,
}: ErrorNotificationProps) => {
    const { t, i18n, ...notificationProps } = props;
    const { data, status } = errorResponse || {};

    const fallbackTitle = i18n.exists(`api:common.error.status.${status}`)
        ? t(`api:common.error.status.${status}`)
        : "There has been an error";

    const title = data?.title || fallbackTitle;
    const message =
        data?.message ||
        t(`api:common.error.${method}.message`, {
            item: props.itemName || "Item",
        });

    notificationService.apiError(message, {
        title,
        message,
        errors: data?.errors,
        ...notificationProps,
    });
};

interface SuccessNotificationProps {
    props: NotificationOptions;
    method: "delete" | "post" | "put";
}

const successNotification = ({ props, method }: SuccessNotificationProps) => {
    const { t, i18n, ...notificationProps } = props;
    const customMessage = `api:${props.localeKey}.success.${method}.message`;
    const shouldOverideMessage = i18n.exists(customMessage);

    const message = shouldOverideMessage
        ? t(customMessage)
        : t(`api:common.success.${method}.message`, {
              item: props.itemName || "Item",
          });

    notificationService.success(message, { ...notificationProps });
};

export { errorNotification, successNotification };
