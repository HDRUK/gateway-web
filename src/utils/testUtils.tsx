import React, { ReactNode } from "react";
import {
    render,
    renderHook,
    RenderHookOptions,
    RenderOptions,
    RenderResult,
    RenderHookResult,
} from "@testing-library/react";
import { SWRConfig } from "swr";
import { CacheProvider, ThemeProvider } from "@emotion/react";
import ActionBarProvider from "@/providers/ActionBarProvider";
import DialogProvider from "@/providers/DialogProvider";
import theme from "@/config/theme";
import createEmotionCache from "@/config/createEmotionCache";

const clientSideEmotionCache = createEmotionCache();

const Wrapper = ({ children }: { children: ReactNode }) => {
    return (
        <SWRConfig
            value={{
                provider: () => new Map(),
            }}>
            <CacheProvider value={clientSideEmotionCache}>
                <ThemeProvider theme={theme}>
                    <ActionBarProvider>
                        <DialogProvider>{children}</DialogProvider>
                    </ActionBarProvider>
                </ThemeProvider>
            </CacheProvider>
        </SWRConfig>
    );
};

interface OptionProps
    extends RenderOptions<
        typeof import("@testing-library/dom/types/queries"),
        HTMLElement,
        HTMLElement
    > {
    wrapperProps: Record<string, unknown>;
}

const customRender = (
    ui: React.ReactElement<
        unknown,
        string | React.JSXElementConstructor<unknown>
    >,
    options?: OptionProps
): RenderResult => {
    const { wrapperProps, ...rest } = options || {};

    return render(ui, {
        wrapper: props => <Wrapper {...props} {...wrapperProps} />,
        ...rest,
    });
};

interface OptionHookProps
    extends RenderHookOptions<
        unknown,
        typeof import("@testing-library/dom/types/queries"),
        HTMLElement,
        HTMLElement
    > {
    wrapperProps: Record<string, unknown>;
}

const customRenderHook = (
    ui: (initialProps: unknown) => unknown,
    options?: OptionHookProps
): RenderHookResult<unknown, unknown> => {
    const { wrapperProps, ...rest } = options || {};
    return renderHook(ui, {
        wrapper: props => <Wrapper {...props} {...wrapperProps} />,
        ...rest,
    });
};

export * from "@testing-library/react";

export { customRenderHook as renderHook };
export { customRender as render };
