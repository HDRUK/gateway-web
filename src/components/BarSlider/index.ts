import BarSlider from "./BarSlider";

export default BarSlider;
