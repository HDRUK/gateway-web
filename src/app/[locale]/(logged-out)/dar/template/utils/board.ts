import { BoardSections, Board, Task } from "../components";
import { BOARD_SECTIONS } from "../constants";
import { getTasksByStatus } from "./tasks";

export const initializeBoard = (tasks: Task[]) => {
    const boardSections: BoardSections = {};

    Object.keys(BOARD_SECTIONS).forEach(boardSectionKey => {
        boardSections[boardSectionKey] = getTasksByStatus(
            tasks,
            boardSectionKey as Board
        );
    });

    return boardSections;
};

export const findBoardSectionContainer = (
    boardSections: BoardSections,
    id: string
) => {
    if (id in boardSections) {
        return id;
    }

    const container = Object.keys(boardSections).find(key =>
        boardSections[key].find(item => item.id === id)
    );
    return container;
};
