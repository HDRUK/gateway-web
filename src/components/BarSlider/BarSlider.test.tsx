import React from "react";
import BarSlider from "@/components/BarSlider";
import { render } from "@/utils/testUtils";

const mockData = [
    { xValue: [0, 1000.0], yValue: 3 },
    { xValue: [1000.0, 2000.0], yValue: 24 },
    { xValue: [2000.0, 3000.0], yValue: 43 },
    { xValue: [3000.0, 4000.0], yValue: 34 },
    { xValue: [4000.0, 5000.0], yValue: 7 },
    { xValue: [5000.0, 6000.0], yValue: 62 },
    { xValue: [6000.0, 7000.0], yValue: 3 },
    { xValue: [7000.0, 8000.0], yValue: 93 },
    { xValue: [8000.0, 9000.0], yValue: 23 },
    { xValue: [9000.0, 10000.0], yValue: 63 },
];

describe("BarSlider", () => {
    it("should render component", async () => {
        const wrapper = render(
            <BarSlider
                step={1000}
                data={mockData}
                ariaLabel="Dataset Population"
                height={130}
            />
        );

        expect(wrapper.container).toMatchSnapshot();
    });
});
