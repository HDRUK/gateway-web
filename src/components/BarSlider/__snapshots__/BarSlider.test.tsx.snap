// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BarSlider should render component 1`] = `
<div>
  <div
    style="height: 130px; margin-bottom: -13px;"
  >
    <div
      style="width: 100%; height: 100%;"
    />
  </div>
  <span
    class="MuiSlider-root MuiSlider-colorPrimary MuiSlider-sizeSmall mui-style-18hov5u-MuiSlider-root"
  >
    <span
      class="MuiSlider-rail mui-style-14pt78w-MuiSlider-rail"
    />
    <span
      class="MuiSlider-track mui-style-1n40zqk-MuiSlider-track"
      style="left: 0%; width: 100%;"
    />
    <span
      class="MuiSlider-mark MuiSlider-markActive mui-style-6cwnna-MuiSlider-mark"
      data-index="0"
      style="left: 0%;"
    />
    <span
      class="MuiSlider-thumb MuiSlider-thumbSizeSmall MuiSlider-thumbColorPrimary MuiSlider-thumb MuiSlider-thumbSizeSmall MuiSlider-thumbColorPrimary mui-style-1s92h5p-MuiSlider-thumb"
      data-index="0"
      style="left: 0%;"
    >
      <input
        aria-label="Dataset Population"
        aria-orientation="horizontal"
        aria-valuemax="10"
        aria-valuemin="0"
        aria-valuenow="0"
        aria-valuetext="0°C"
        data-index="0"
        max="10"
        min="0"
        step="1000"
        style="border: 0px; height: 100%; margin: -1px; overflow: hidden; padding: 0px; position: absolute; white-space: nowrap; width: 100%; direction: ltr;"
        type="range"
        value="0"
      />
      <span
        aria-hidden="true"
        class="MuiSlider-valueLabel mui-style-ytpj86-MuiSlider-valueLabel"
      >
        <span
          class="MuiSlider-valueLabelCircle"
        >
          <span
            class="MuiSlider-valueLabelLabel"
          >
            1
          </span>
        </span>
      </span>
    </span>
    <span
      class="MuiSlider-thumb MuiSlider-thumbSizeSmall MuiSlider-thumbColorPrimary MuiSlider-thumb MuiSlider-thumbSizeSmall MuiSlider-thumbColorPrimary mui-style-1s92h5p-MuiSlider-thumb"
      data-index="1"
      style="left: 100%;"
    >
      <input
        aria-label="Dataset Population"
        aria-orientation="horizontal"
        aria-valuemax="10"
        aria-valuemin="0"
        aria-valuenow="10"
        aria-valuetext="10°C"
        data-index="1"
        max="10"
        min="0"
        step="1000"
        style="border: 0px; height: 100%; margin: -1px; overflow: hidden; padding: 0px; position: absolute; white-space: nowrap; width: 100%; direction: ltr;"
        type="range"
        value="10"
      />
      <span
        aria-hidden="true"
        class="MuiSlider-valueLabel mui-style-ytpj86-MuiSlider-valueLabel"
      >
        <span
          class="MuiSlider-valueLabelCircle"
        >
          <span
            class="MuiSlider-valueLabelLabel"
          >
            10000
          </span>
        </span>
      </span>
    </span>
  </span>
</div>
`;
